Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,265 @@
+            // since we want to signal them individually and wait for a single service
+            // to finish before moving to the next one
+            var gracefulShutdownManagers = [GracefulShutdownManager]()
+            gracefulShutdownManagers.reserveCapacity(self.services.count)
+
+            for (index, service) in self.services.enumerated() {
+                self.logger.debug(
+                    "Starting service",
+                    metadata: [
+                        self.configuration.logging.serviceKey: "\(service)",
+                    ]
+                )
+
+                let gracefulShutdownManager = GracefulShutdownManager()
+                gracefulShutdownManagers.append(gracefulShutdownManager)
+
+                group.addTask {
+                    return await TaskLocals.$gracefulShutdownManager.withValue(gracefulShutdownManager) {
+                        do {
+                            try await service.run()
+                            return .serviceFinished(service: service, index: index)
+                        } catch {
+                            return .serviceThrew(service: service, index: index, error: error)
+                        }
+                    }
+                }
+            }
+
+            precondition(gracefulShutdownManagers.count == self.services.count, "We did not create a graceful shutdown manager per service")
+
+            // We are going to wait for any of the services to finish or
+            // the signal sequence to throw an error.
+            while !group.isEmpty {
+                // No child task is actually throwing here so the try! is safe
+                let result: ChildTaskResult? = await group.next()
+
+                switch result {
+                case .serviceFinished(let service, _):
+                    // If a long running service finishes early we treat this as an unexpected
+                    // early exit and have to cancel the rest of the services.
+                    self.logger.error(
+                        "Service finished unexpectedly. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                        ]
+                    )
+
+                    group.cancelAll()
+                    return .failure(ServiceRunnerError.serviceFinishedUnexpectedly())
+
+                case .serviceThrew(let service, _, let error):
+                    // One of the servers threw an error. We have to cancel everything else now.
+                    self.logger.error(
+                        "Service threw error. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                            self.configuration.logging.errorKey: "\(error)",
+                        ]
+                    )
+                    group.cancelAll()
```

Comment:
Do we always want that? Especially the cancel? We might have unimportant services that, if they can't do their job anymore, we want to gracefully shutdown, but not cancel stuff that is in flight?

Reply:
I was just thinking about this as well. For example if a background credentials rotation that fails, it is likely we have some time (we rotate credentials 5 minutes before they are due to expire), before its going to become an issue. Likely what we would like to do is stop responding successfully to health checks so the instance stops receiving new requests and then drops out of the load balancer but survives long enough to successfully complete anything that is in flight.

Reply:
Thanks for bringing up a concrete example!

I don't think this requires any extra work in this library to support. Your service that is doing the credential refresh should expose its state somehow. Your health checking code will then check that state on every health check request and fail it when the scenario above happens.
Furthermore, the service `run()` can itself handle this by `Task.sleep`ing after the crendetial refresh failed for an appropriate amount of time. This is in my opinion very service logic specific since you need to see how long your credentials are still valid when the fresh failed and sleep for an appropriate amount of time.

In general, this goes to show that these "short"/"not essential" running services can be modelled with the current APIs. While they might require some extra work in the `run()` method they are possible to implement. This allows us to defer implementing something in this library and wait how adoption turns out.

Reply:
This seems reasonable. It is essentially saying the correct way to shutdown an running application is for a signal to be thrown and any service that decides it is not in a good state should keep running but have some alternate mechanism to cause that signal. It certainly seems the correct behaviour for not-immediately-essential services in our use case; they have to go through the health check mechanism to protect the requests being processed by the critical service. I don't have any counter examples at the moment.

The work potentially is - completely outside of this - to define a common health check mechanism that makes this sort of signalling easy, if we consider that useful. That would probably fall more on the middleware side of things.

Reply:
Health checking is definitely something interesting to consider (though agree outside of this). In my opinion, a good pattern is for each concrete `Service` to expose their external state e.g. `running` or `shutdown` through a property or an async sequence. This enables to implement a health check which listens to all the states and responds correctly.
In most orchestration systems repeated health check failures lead to `SIGTERM` or `SIGKILL` so this is going to end up shutting down the application.